### DISCUSSION TITLE
What would you like to see?

### DISCUSSION THREAD

**Discussion by wawawario2** [wawawario2/text-generation-webui#5] [General]
Feel free to spitball any features you'd like to see. Here are some ideas on my backburner:
- Support multiple characters seamlessly
- Look into using langchain (with perhaps chroma) for the backend. If I do this, I'll write a script to migrate your memories seamlessly.
- Have memories stick around for more than just one message
- Use a LLM to summarize conversations and use summaries for memories (I think this has a lot of potential of removing filler text from memories)

**Comment by JoureGameDev** [comment]
I can't come up with anything since I just got it but yeah, multiple characters would be my number one request! That and having memories stick around!

**Comment by phiferch** [comment]
So, I've been thinking about this a lot lately and forgive me if it makes no sense. I'm a software engineer, but I've never done any work with LLMs before. The context window is just a box for all the relevant information necessary to formulate an appropriate continuation of the text. We treat it like its short-term memory, and your database acts like long term memory, injecting text into the context window. You intentionally try not to flood the context window, as the memories consist of whole raw conversations. 

Here's my thought, and again if it makes no sense, just say so. Let's say memories don't consist of raw conversations in their entirety, but rather individual replies from both the AI and the User.  Then you partition the context window into 3 sections, "Permanent Context" (e.g. personality of the AI, and the User), "Relevant Context" (e.g. injected long term memories), and "Immediate Context" (e.g. the last reply from the user). Then, instead of trying not to fill the context window, we did the opposite filling the "Relevant Context" portion of the window with as many relevant memories as you can fit, selecting relevant memories not just by their cosine-similarity, but by their age, ordering them linearly. Older memories having less relevance than newer ones. Then continuing the text generation from the "Immediate Context" portion of the window. 

Instead of a scrolling context window, would you have a sort of well informed, one-shot text generating AI?

Edit to add some thoughts:
I realize, this is very similar to the way you do this already, at least as much as I understand it, but what I was trying to think of was a way to get away from holding the entire conversation in the context window, and instead having it just try to formulate one individual response at a time, using nothing but what it must always know, combined with what we tell it is relevant to formulate a convincing response. I'm not sure if this does that, but I would be interested in anyones thoughts on it.

**Comment by snicolast** [comment]
I do not know how similar this project is with this: https://memorygpt.io/about-project.html That technique seems to work quite well. Maybe it could be good to give it a try and see if it's possible to replicate?

I also vote for langchain and chroma support, apart from support to multiple characters. Keep it up.
